src/core/main/Locking/InMemoryDistributedAppLock.cs

```diff
@@ -0,0 +1,165 @@
+using System;
+using System.Collections.Concurrent;
+using System.Diagnostics;
+using System.Threading.Tasks;
+
+namespace RapidCore.Locking
+{
+    public class InMemoryDistributedAppLock : IDistributedAppLock
+    {
+        private bool _disposedValue; // To detect redundant calls
+        private readonly Random _rng;
+        private readonly ConcurrentDictionary<string, InMemoryDistributedAppLockSemaphore> _semaphores;
+
+        public InMemoryDistributedAppLock(Random rng, ConcurrentDictionary<string, InMemoryDistributedAppLockSemaphore> semaphores)
+        {
+            _rng = rng;
+            _semaphores = semaphores;
+        }
+
+        /// <summary>
+        /// Acquire the lock
+        /// </summary>
+        /// <param name="lockName">Name of the lock to acquire</param>
+        /// <param name="lockWaitTimeout">When set, the amount of time to wait for the lock to become available</param>
+        public async Task<IDistributedAppLock> AcquireLockAsync(string lockName, TimeSpan? lockWaitTimeout = null)
+        {
+            WasAcquiredInstantly = true;
+            var timeoutProvided = lockWaitTimeout.HasValue;
+            if (!timeoutProvided)
+            {
+                lockWaitTimeout = TimeSpan.Zero;
+            }
+            
+            InMemoryDistributedAppLockSemaphore semaphore;
+            lock (string.Intern(lockName))
+            {
+                semaphore = _semaphores.GetOrAdd(lockName, _ => new InMemoryDistributedAppLockSemaphore());
+                semaphore.IncrementReferenceCount(); // Remember that this lock is being referenced
+            }
+            
+            var timeout = TimeSpan.Zero;
+            
+            var stopWatch = new Stopwatch();
+            do
+            {
+                var lockWasAcquired = await semaphore.WaitAsync(timeout);
+                
+                if (!lockWasAcquired && !timeoutProvided)
+                {
+                    LockRelease(lockName);
+                    throw new DistributedAppLockException("Unable to acquire lock")
+                    {
+                        Reason = DistributedAppLockExceptionReason.LockAlreadyAcquired,
+                    };
+                }
+
+                if (!lockWasAcquired)
+                {
+                    WasAcquiredInstantly = false;
+                    timeout = TimeSpan.FromMilliseconds(_rng.Next(
```

Comment:
After closer inspection, I don't think this random timeout is necessary.

I based this implementation on the `Redis` implementation, where polling is necessary.
But I would expect the semaphore to make this obsolete.

Reply:
No it shouldn't be necessary, we use the random polling interval in the redis provider to avoid hammering the Redis server, see this issue: https://github.com/rapidcore/rapidcore/issues/147